[FEAT] Add custom CUDA `tinygemm` unpacker

jeromeku opened the pull request:
## Description

Adds CUDA custom ops to unpack weights that have been packed with `torch.ops.aten._convert_weight_to_int4pack` for use with `torch.ops.aten._weight_int4pack_mm`.  

Currently there is only a packing function that permutes and prepacks the weights in tensor-core format.  However, there is no equivalent unpacking function that reorders the weights back to the original logical layout.

The implementation is an adaptation of the original packing code (`int4mm.cu`) with modifications to simplify indexing logic and fused unpacking & dequantization.

## Motivation
Fast unpacking of packed weights is needed when switching quantized gemm backends during inference.  

As workloads transition from memory-bound to compute-bound (i.e., context length growth during decoding), users might wish to switch to a different kernel implementation that is more performant in this regime than `tinygemm`.  

In order to do this, the weights need to be unpacked from the packed format.  Alternative would be to store 2 copies of the weights -- one packed, one in logical format -- but this is clearly not ideal given memory burden.

## Features
Add 2 custom CUDA ops, registered per the instructions in `torchao` custom op documentation:
1) `torchao.ops.unpack_int4` - unpacks the packed weight to the original `N x K` logical layout with dtype `torch.int`.  Can be used within  [TensorCoreTiledAQTLayout.get_plain](https://github.com/pytorch/ao/blob/9dc2c118f59ad4135a8c39166c4ceebda73c62a9/torchao/dtypes/affine_quantized_tensor.py#L505) to recover original layout of the (quantized) tensor. 
2) `torchao.ops.dequantize_int4` - dequantizes the packed weight to `bfloat16` tensor with original `N x K` logical layout.  This is useful for developers who want to unpack and dequantize the packed weight when switching quantized matmul backends on the fly.
 
## Tests
Tests have been added to `test/test_ops.py` for both correctness as well as for correct custom op registration.  

> Note: `opcheck` `test_aot_dispatch_dynamic` currently failing, investigating.

## TODO
- [x] Fuse dequant into unpacking kernel
  - ~~Kernel works against a reference implementation per my understanding of dequantization but needs further verification (see notes in `test/test_ops.py:test_dequant_int4_correctness`)~~
- [x] Implement `dequantize` for `ZeroPointDomain.Float` per [update](https://github.com/pytorch/ao/pull/415#issuecomment-2187016445)
- [x] Debug `test_aot_dispatch_dynamic` `opcheck` failure 
- [ ] Integrate with `AQT` `get_plain`

@msaroufim

jerryzh168 commented:
great, after this is landed we can replace this workaround code https://github.com/pytorch/ao/blob/9dc2c118f59ad4135a8c39166c4ceebda73c62a9/torchao/dtypes/affine_quantized_tensor.py#L505 with the op I think

jeromeku commented:
@jerryzh168 

Took a look at `get_plain` for int4 `AQT` type: will revise the kernel so that it more closely aligns with the API of `get_plain`.

jerryzh168 commented:
> @jerryzh168
> 
> Took a look at `get_plain` for int4 `AQT` type: will revise the kernel so that it more closely aligns with the API of `get_plain`.
> 
> Are there tests for `get_plain` that I can adapt to verify my implementation?
> 
> Currently working on fusing dequant into the unpacking kernel, however, a simple [sanity check](https://gist.github.com/jeromeku/0ee8e31a44733ba69930f4a0a3df8fc2) using the same logic as `get_plain` is failing.
> 
> That is, I'm using `tinygemm` to dequantize by passing in an identity matrix as operand `a` and packed weights, scales, and zeros, respectively. Comparing against a reference dequantize method that does `(q - zero) * scale` on the unpacked weights, scales, and zeros does not check out. Am I misinterpreting the scales and zeros?

we don't have get_plain() tests yet, but I'm planning to add some tests for AffineQuantizedTensor in the future

the way that tinygemm dequantize implmeneted is a bit different from the normal path, here is how it's implmeneted:

function call to our primitive ops: https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/utils.py#L370
code path: https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L266

main difference is the zero_point is in floating point domain (while the the quant/dequant that we are more familiar with is in integer domain): https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L43-L44

jeromeku commented:
@jerryzh168 

Many thanks on the clarification!  

Helps explain why `tinygemm` kernel is able to use a single [fma](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L467-L475) to dequantize (using `integer` zero-point would require a `sub` then `mul` unless zeros are stored as scales * zeros, which is not the case).  

This is good to know as the original motivation for this PR was to help `answer.ai / hqq` developers who are using `tinygemm` as a quantized matmul backend.  However, I believe `hqq` is using the `integer` zeropoint derivation (but keeping the zeropoint in the original floating-point dtype), which will result in incorrect results when using `tinygemm` kernel, which is dequantizing based on the `floating-point` zeropoint calculation.

What is the mathematical derivation of the `float` dequantization method vs the more common `integer` quantization scheme?  Are there any papers / blogs that explain the reasoning for this difference?

jerryzh168 commented:
> Helps explain why `tinygemm` kernel is able to use a single [fma](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L467-L475) to dequantize (using `integer` zero-point would require a `sub` then `mul` unless zeros are stored as scales * zeros, which is not the case).

yes, the motivation for `tinygemm` to have zero_point in floating point domain is exactly to use a single `fma` (I talked to Jeff about this).

> This is good to know as the original motivation for this PR was to help `answer.ai / hqq` developers who are using `tinygemm` as a quantized matmul backend. However, I believe `hqq` is using the `integer` zeropoint derivation (but keeping the zeropoint in the original floating-point dtype), which will result in incorrect results when using `tinygemm` kernel, which is dequantizing based on the `floating-point` zeropoint calculation.

for `hqq` yeah we need to make sure this detail is correct since they are using tinygemm kernels. cc @HDCharles @mobicham 


> What is the mathematical derivation of the `float` dequantization method vs the more common `integer` quantization scheme? Are there any papers / blogs that explain the reasoning for this difference?

I'm not aware of any formal papers or blogs. so the differences are shown in our quant_primitive ops in these two flags: https://github.com/pytorch/ao/blob/c2cf9730ae695e83e27cbb304d56fe1f19a06659/torchao/quantization/quant_primitives.py#L303-L318

traditional integer quantization:
1. preserve_zero is True: this is because traditionally we use quantization on conv, and it can have zero_padding, so there is a domain specific requirement of floating point zero has to be exactly representable: https://github.com/google/gemmlowp/blob/master/doc/quantization.md#domain-specific-constraint-the-real-value-0-must-be-exactly-representable

2. zero_point is in integer domain
This is probably for static quantization where there are hardwares that only support integer compute

tinygemm:
1. preserve_zero is False because mainly we care about linear, this will also help improve accuracy in some cases since we don't need to always include zero during quantization
2. zero_point is in floating point domain
this is because of `fma` I think

jeromeku commented:
@jerryzh168 

Thanks!

Looking at `gpt-fast` quantization code [here](https://github.com/pytorch-labs/gpt-fast/blob/e71d268012881c619f049f39626df1cf3510e186/quantize.py#L58-L76), scales / zeros are calculated as: 
```python
  scales = (max_val - min_val).clamp(min=1e-6) / max_int
  zeros = min_val + scales * (2 ** (n_bit - 1)) 
```
where `2 ** (nbit-1)` is the mid-point.  Hence it is fusing the mid-point into the zeros.

Then in the tinygemm kernel, [dequantization](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L467-L475) is performed as `q * scale + zero`, which is close to but does not match exactly `ZeroPointDomain.FLOAT` [dequant](https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L43-L44).  Can you elaborate on what `zero_point (float)` refers to [here](https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L44)?

jerryzh168 commented on torchao/csrc/cuda/unpack_int4/unpack_int4.cu:
```diff
@@ -0,0 +1,268 @@
+}
+
+TORCH_LIBRARY_IMPL(torchao, CUDA, m) {
+  m.impl("torchao::unpack_int4_to_int", &_unpack_int4_to_int);
+  m.impl("torchao::dequantize_int4", &_dequantize_int4);
```
nit: I feel we probably need to mention `tensor_core_tiled` layout in the name of these ops if these are specific to that packing format

jerryzh168 commented on torchao/csrc/unpack_int4.cpp:
```diff
@@ -0,0 +1,10 @@
+TORCH_LIBRARY_FRAGMENT(torchao, m) {
+  m.impl_abstract_pystub("torchao.ops");
+  m.def("unpack_int4_to_int(Tensor packed_w, int innerKTiles) -> Tensor");
+  m.def("dequantize_int4(Tensor packed_w, Tensor scales_and_zeros, int group_size, int innerKTiles) -> Tensor");
```
so this op is actually dequant + unpack?

jerryzh168 commented:
> @jerryzh168
> 
> Thanks!
> 
> Looking at `gpt-fast` quantization code [here](https://github.com/pytorch-labs/gpt-fast/blob/e71d268012881c619f049f39626df1cf3510e186/quantize.py#L58-L76), scales / zeros are calculated as:
> 
> ```python
>   scales = (max_val - min_val).clamp(min=1e-6) / max_int
>   zeros = min_val + scales * (2 ** (n_bit - 1)) 
> ```
> 
> where `2 ** (nbit-1)` is the mid-point. Hence it is fusing the mid-point into the zeros.
> 
> Then in the tinygemm kernel, [dequantization](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L467-L475) is performed as `q * scale + zero`, which is close to but does not match exactly `ZeroPointDomain.FLOAT` [dequant](https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L43-L44). Can you elaborate on what `zero_point (float)` refers to [here](https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L44)?

sorry I just wrote down the quantize function there, it's not  the dequant function, I should probably add all algorithms (choose_qparams, quant, dequant) there. the dequant we are using is here: https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L267-L274

jerryzh168 commented:
> @jerryzh168 Sorry for persisting on this matter but still a gap in my understanding:
> 
> If we unpack what `gpt-fast` and `tinygemm` is doing:
> 
> ```
>   scales = (max_val - min_val).clamp(min=1e-6) / max_int
>   zeros = min_val + scales * (2 ** (n_bit - 1)) = min_val + scales * mid_point
> ```
> 
> Then in `tinygemm`, dequantization is calculated as:
> 
> ```python
> x = q * scales + zeros 
>    =  q * scales + min_val + scales * mid_point
> ```
> 
> Where `x` is dequantized value, `q` is quantized value.
> 
> Comparing this to `torchao` dequant per your link:
> 
> ```python
>  x = (q - mid_point) * scales + zeros 
>     = q * scales - scales * mid_point + zeros
> ```
> 
> Assuming `zero_point` is calculated per `gpt_fast`:
> 
> ```python
> x = q * scales - scales * mid_point + min_val + scales * mid_point
>    = q * scales + min_val
> ```
> 
> How to reconcile these differences? How are `zeros` expected to be calculated in `torchao`?

yeah no problem, I think the main difference as you listed is this part:

```
> Then in `tinygemm`, dequantization is calculated as:
> 
> ```python
> x = q * scales + zeros 
>    =  q * scales + min_val + scales * mid_point
> ```
```

I feel the `q` here should probably be `(q - mid_point)`

I'm not very familiar with tinygemm kernel implementation itself, but I think this should be accounted for either by preprocess of `q` or post process of the results after that dequant op.

also for some additional context, current quant primitives in torchao are adapted from the original gpt-fast/tinygemm choose_qparams/quantize/dequantize implemnetations and we have regression tests to make sure they match: https://github.com/pytorch/ao/blob/c2cf9730ae695e83e27cbb304d56fe1f19a06659/test/quantization/test_quant_primitives.py#L49

jerryzh168 commented:
maybe related to https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L197

cc @HDCharles do you know how tinygemm kernel dequant implementation match up with the python dequant implementation?

jeromeku commented:
@jerryzh168 

For the purposes of this PR, then, what should be the expected behavior of `dequantize_int4`?

That is, given packed weights, scales, zeros, etc., what should be the calculation to dequantize the weights from `int4` to `bfloat16`?

Checking the [quant_primitives](https://github.com/pytorch/ao/blob/37c348e0a6f33398243f5a74adc86c5729d87888/torchao/quantization/quant_primitives.py#L267-L274) `dequant` methodology against calling `tinygemm` with an identity matrix to dequant gives a small error (~ 1e-2), see this [script](https://gist.github.com/jeromeku/98f318198d3de21d7916d379d8e949f5).

jerryzh168 commented:
@jeromeku OK I just confirmed with Jeff Johnson that [this code](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/int4mm.cu#L197) is actually doing both a uint4 -> int4 conversion ([0, 15] --> [-8, 7]) which is equivalent to (q_val - mid_point) in our dequant code, and also a conversion to bfloat16

so I feel the `dequantize_op` in this case should follow what we are doing in `dequantize_affine` op, at least for int4, I also need to think a bit about this `uint4 -> int4` conversion stuff, I feel it should probably be done outside of quant primitives op

for test, what you described make sense, I think you can do two test:
1. quant + dequant v.s. (quant + pack + packed_dequant_int4 (your op))
2. quant + pack + tinygemm_int4_mm with identity matrix v.s. (quant + pack + packed_dequant_int4 (your op))

jeromeku commented:
@jerryzh168 

Thanks for the clarification.

Will update the `dequant` kernel to reflect this change: `u4 -> s4 + upcast` followed by `scale + shift` and add relevant tests. 

Would be good to add some additional documentation explaining the pre-processing / post-processing needed to use quantized weights, scales, and zero-points prepared using "conventional" (`ZeroPointDomain.INT`) schemes for use with `tinygemm`, e.g., `hqq`.

jerryzh168 commented:
sure thanks, I'll add some docs for quant_primitives in our README

jeromeku commented:
@jerryzh168 

Updates:
- Updated the `dequantize` kernel to match that of `ZeroPointDomain.FLOAT` methodology
- Correctness tests:
  - The `dequant` kernel results match those of the "identity matrix" `tinygemm` hack exactly
  - Both the `dequant` kernel and the identity matrix `tinygemm` have numerical differences ~`1e-2` when compared to `group_affine_dequantize`, likely due to the bit twiddling trick used to convert `int` to `bf16`.
- Renamed the `unpack` and `dequantize` kernels to `unpack_tensor_core_tiled_layout` and `dequantize_tensor_core_tiled_layout` to align with `torchao.quant` API naming conventions

jerryzh168 commented:
Thanks @jeromeku, I'll take a look next week

jerryzh168 commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(IS_FBCODE, reason="Skipping the test in fbcode since we don't have TARGET file for kernels")
+@pytest.mark.parametrize("shape, innerKTiles, group_size", TEST_CONFIGS_DEQUANT, ids=str)
+def test_dequantize_tensor_core_tiled_layout_correctness(shape, innerKTiles, group_size):
```
nit: maybe we can use `inner_k_tiles` instead of `innerKTiles` to align the naming convention for ags

jerryzh168 commented on torchao/csrc/cuda/tensor_core_tiled_layout/tensor_core_tiled_layout.cu:
```diff
@@ -0,0 +1,312 @@
+// input is [n / 8][k / (InnerKTiles * 16)][32][innerKTiles / 2]
+// scales_and_zeros is [numQGroups][n][2]
+// qGroupSize is 32, 64, 128 or 256
+at::Tensor _dequantize_tensor_core_tiled_layout(
```
one additional high level thought for tensor_core_tiled_layout, is that, can we make this a layout for `uint4` tensor in the future? or is this tied to the affine quantization logic? it feels like we can move this over to be a layout for `uint4`

jerryzh168 commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+    t = torch.randint(0, 16, dtype=torch.int, size=shape, device="cuda")
+    packed_w = torch.ops.aten._convert_weight_to_int4pack(t, innerKTiles)
+
+    opcheck(
```
which pytorch version are you using? it seems this `opcheck` is moved to `torch.library.opcheck`: https://github.com/pytorch/pytorch/pull/124496/files

jeromeku replied:
torch `2.5.0.dev20240624+cu121`

jerryzh168 commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+    t = torch.randint(0, 16, dtype=torch.int, size=shape, device="cuda")
+    packed_w = torch.ops.aten._convert_weight_to_int4pack(t, innerKTiles)
+    unpacked = torchao.ops.unpack_tensor_core_tiled_layout(packed_w, innerKTiles)
+    assert torch.allclose(t, unpacked)
```
should this be equal?

jerryzh168 commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+    diff_op_ao = (dq_op - dq_ao).abs().max()
+    
+    # There are slight numerical differences when dequantizing with an identity matrix
+    # Since the `dequantize_int4` kernel relies on same underlying numerical conversions, this gives same
```
nit: kernel name need to be updated

jerryzh168 commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+    # numerical differences when compared to the `groupwise_affine_dequantize`
+    
+    # Test that the `dequant` kernel gives same results as identity matrix-based dequant 
+    assert diff_op_id == 0
+    
+    # Test that the `dequant` kernel gives same numerical diffs as the `groupwise_affine_dequantize` when compared against the identity matrix
+    assert diff_op_ao == diff_ao_id
```
this doesn't seem to match the description, I thought you were saying "dequant with id matrix" is a bit off, and diff_op_ao == 0, so it should be:

assert diff_op_ao == 0
assert diff_op_id == diff_ao_id

?

jeromeku replied:
I'm using the same bit twiddling trick used in the `tinygemm` fused matmul that does a fast conversion and upcast of `u4 -> s4 -> bf16`.  The results of the the fused dequant kernel match those exactly of the id matrix tinygemm hack, but both of these have numerical differences (~1e-2) with native torch-native (`ao`) dequant.

jerryzh168 replied:
OK sounds good, I think you updated the doc, thanks

jerryzh168 commented on torchao/ops.py:
```diff
@@ -43,3 +43,108 @@ def _(_in_feats, _weights, _scales, splitK = 1):
+    N = packed_w.size(0) * 8
+    K = packed_w.size(1) * innerKTiles * 16
+
+    return torch.empty((N, K), dtype=torch.int32, device=packed_w.device)
```
is this supposed to call `unpack_tensor_core_tiled_layout`?

jerryzh168 commented on torchao/ops.py:
```diff
@@ -43,3 +43,108 @@ def _(_in_feats, _weights, _scales, splitK = 1):
+    torch._check(scales_and_zeros.size(1) == N, lambda: "scales_and_zeros must have N at dim 1")
+    torch._check(scales_and_zeros.size(2) == 2, lambda: "scales_and_zeros must have 2 at dim 2")
+    
+    return torch.empty((N, K), dtype=torch.bfloat16, device=packed_w.device)
```
same for this, is this supposed to call `dequantize_tensor_core_tiled_layout`

jeromeku replied:
I thought this was the expected pattern for registering a custom op?  I was following the example of the pre-existing `fp6_linear` custom op already in `ops.py`.

Previously one would register an `abstract impl` for composability with `torch.compile`.  Thought this was the expected interface with the new custom op registration API.

jerryzh168 replied:
oh OK, I think I understand now, `register_custom_op` is calling `register_fake`/`impl_abstract`, I feel we need to rename this util to something more accurate, cc @msaroufim

jerryzh168 commented on torchao/ops.py:
```diff
@@ -43,3 +43,108 @@ def _(_in_feats, _weights, _scales, splitK = 1):
+
+    return torch.empty((N, K), dtype=torch.int32, device=packed_w.device)
+
+def dequantize_tensor_core_tiled_layout(packed_w: Tensor, scales_and_zeros: Tensor, group_size: int, innerKTiles: int) -> Tensor:
```
do we really need two ops btw, can we just have an unpack op and then call dequantize afterwards to achieve the same thing?

jeromeku replied:
Yes - for performance since the second fuses unpacking from tensor-core to logical layout and then dequantizes.

jeromeku commented on test/test_ops.py:
```diff
@@ -55,6 +64,162 @@ def test_fp6_llm_linear_correctness(self, BS, OC, IC, splitK):
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(IS_FBCODE, reason="Skipping the test in fbcode since we don't have TARGET file for kernels")
+@pytest.mark.parametrize("shape, innerKTiles, group_size", TEST_CONFIGS_DEQUANT, ids=str)
+def test_dequantize_tensor_core_tiled_layout_op(shape, innerKTiles, group_size):
```
Added the first case.  I thought the second is already tested by the existing correctness test.

jerryzh168 replied:
oh OK yeah makes sense

jeromeku commented:
@jerryzh168 

Fixed all the above:
- [x] Renamed `innerKTiles` -> `inner_k_tiles`
- [x] Changed `unpack` test from testing for closeness to equality
- [x] Added additional `(unpack_tensor_core_tiled_layout_op + dequant) vs. dequantize_tensor_core_tiled_layout_op` test
- [x] Added comments clarifying the logic of the fused `dequant` kernel tests
  - Since `tinygemm` id matrix dequant hack and the `fused dequant` kernel utilize the same underlying fast `CUDA` numeric conversion path from `u4 -> s4 -> bf16`, they have identical numerics  
  - Both result in ~1e-2 discrepancies when compared with `ao` `groupwise_affine_dequantize`
  - These conditions are tested for in the `dequantize_tensor_core_layout` tests:
    - difference between `tinygemm id matrix dequant` and `dequant_tensor_core_layout` is 0
    - the difference between `tinygemm id matrix` vs. `groupwise_affine_dequant` is the same as the difference between `dequantize_tensor_core_layout` and `groupwise_affine_dequant`.

jerryzh168 commented on torchao/ops.py:
```diff
@@ -43,3 +44,108 @@ def _(_in_feats, _weights, _scales, splitK = 1):
+
+    return torch.empty((N, K), dtype=torch.int32, device=packed_w.device)
+
+def dequantize_tensor_core_tiled_layout(packed_w: Tensor, scales_and_zeros: Tensor, group_size: int, inner_k_tiles: int) -> Tensor:
```
is this specific to `int4` btw?

jerryzh168 reviewed (approved):
LGTM! really appreciate adding this functionality and the thorough comments/testing!

msaroufim commented:
Just a minor merge conflict and this should be good to merge

jeromeku commented:
@jerryzh168 @msaroufim 

Getting CI failure unrelated to PR:
```
 =========================== short test summary info ============================
  FAILED test/integration/test_integration.py::SmoothquantIntegrationTest::test_on_dummy_distilbert - requests.exceptions.ReadTimeout: (ReadTimeoutError("HTTPSConnectionPool(host='huggingface.co', port=443): Read timed out. (read timeout=10)"), '(Request ID: 72162155-ccbe-44a5-9304-e6e08c061a4e)')
  ====== 1 failed, 202 passed, 556 skipped, 11 warnings in 93.47s (0:01:33) ======
  Error: Process completed with exit code 1.
```

msaroufim reviewed (approved):
Cool thank you for the awesome work @jeromeku and thank you for the thorough review @jerryzh168 

The CI failure indeed seems unrelated, most likely a flake due to connection issues with HF
